Features/aperta 7046 content dropped on transform

MMercieca opened the pull request:
`v:textbox` elements are always contained (or at least it seems so) inside of `w:pict` elements.  We throw out `w:pict` elements so we never see these textboxes.  This PR whitelists some elements inside of `w:pict` elements for later processing.

![2016-07-26_15-27-44](https://cloud.githubusercontent.com/assets/2379553/17154079/80e92484-534d-11e6-9bd2-49690211503f.png)

MMercieca committed:
APERTA-7046 Removing mc:Choice nodes and always using the fallback

MMercieca committed:
AERTA-7046 Indenting fixes

MMercieca committed:
APERTA-7046 Selecting textboxes out of w:pict elements

MMercieca committed:
APERTA-7046 Only grabbing textbox children of pict

MMercieca committed:
APERTA-7046 Grabbing content of textboxes that we can process.

Presenting inline message for what we can't.

Bestra commented on docx/from/docxtotei.xsl:
```diff
@@ -1,713 +1,743 @@
+          select="$styleDoc"/> cannot be read</xsl:message>
+    </xsl:if>
+
+    <xsl:variable name="passOneHalf">
```
Could we rename this to pass0 and increment the other passes to 1 and 2 respectively? I think this would help maintain naming consistency.

MMercieca replied:
Yes, we could rename them.  But I like keeping the consistency in the names between these stylesheets and the source. It'll make it easier to port fixes over from the base.

Though I suppose I blew that away when I formatted the document.  Humm.

MMercieca replied:
I renamed our pass to `preprocessing`. I want to keep `pass0` and `pass2` to match the source.

Bestra commented on docx/from/docxtotei.xsl:
```diff
@@ -1,713 +1,743 @@
+    </xsl:template>
+
+    <xsl:template name="removeNestedTables" match="w:txbxContent/w:tbl" mode="passOneHalf">
+      Could not process table here.  Table was removed.
```
do you want to wrap this in any kind of element to help style it later? The thought was maybe this needs to be called out as a warning.

MMercieca replied:
Um, maybe?  I'll give it a whirl.
- [x] Wrap error in something.

MMercieca replied:
So I can wrap the error in a paragraph.  Because of the step, the wrapping has to be in Word XML format.  I was able to reverse engineer the paragraph though.

Bestra commented on docx/from/docxtotei.xsl:
```diff
@@ -1,713 +1,743 @@
+
+    <!-- We cannot process pict directly, but we can try to process textboxes inside them. -->
+    <xsl:template name="unnestFromPict" match="w:pict" mode="passOneHalf">
+      <xsl:apply-templates  select="v:shape/v:textbox/w:txbxContent/*" mode="passOneHalf" />
```
there are two spaces before `select=` here

MMercieca replied:
- [x] Fix select formatting

Bestra commented on docx/from/docxtotei.xsl:
```diff
@@ -1,713 +1,743 @@
+    </xsl:template>
+
+
+    <xsl:template match="/ | @* | node()" mode="passOneHalf">
```
I'd love an explanation of how this is working (either here the comments, or in chat or SH later. Does the `unestFromPict` template apply this template with its `apply-templates` call? If you call `apply-templates mode="foo"` from a `mode="foo"` template, does _that_ template get run again too?

MMercieca replied:
I'll chat or Screenhero any time. :smile:

I'll give it a try here though.  Templates are applied from the most specific match to the least specific match. But once a match is made, no future templates are applied.  

This is what caused the problem we saw where the raw Word XML was making it in to the final document.  We grabbed all of the content we wanted, but since we didn't do it in an `apply-templates` call, all processing stopped.

That's also why we need the identity transform in this pass.  If a node doesn't match any of the selectors in a pass, it won't be selected and passed on.  The identity selection is the least specific call, so it runs last and makes sure we get the whole document.

And yes, if you call `apply-templates mode="foo"` from an `apply-templates mode="foo"` that template will get applied again to all content that matches the selectors.  This is what we want to happen, but can lead to some pretty deep recursion when applied improperly.  It's a consideration for development, and I did see a stackoverflow once or twice, but once the selectors are set up properly we should be good.

Bestra commented on docx/from/docxtotei.xsl:
```diff
@@ -1,713 +1,743 @@
-    <!-- We cannot process these elements. -->
+    <!-- We cannot process pict directly, but we can try to process textboxes inside them. -->
     <xsl:template match="w:pict">
+      <xsl:copy-of select=".//v:textbox"/>
```
is this `copy-of` made redundant by the pass0 operations?

Bestra commented on docx/from/docxtotei.xsl:
```diff
@@ -1,713 +1,743 @@
     </xsl:template>
-    
 
+    <xsl:template match="mc:Choice">
```
could you move this to pass0 as well for consistency?

MMercieca committed:
APERTA-7046 Updates from code review

- Renamed passOneHalf to preprocessing
- Better display of table removal error
- Formatting fix

Bestra commented:
Good explanation! I have one final request.

MMercieca committed:
APERTA-7046 Making sure all new templates are applied in the correct pass.

egh commented on docx/from/docxtotei.xsl:
```diff
@@ -1,713 +1,750 @@
+  xmlns:ve="http://schemas.openxmlformats.org/markup-compatibility/2006"
+  xmlns:w10="urn:schemas-microsoft-com:office:word"
+  xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main"
+  xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
```
This namespace already exists as `ve` (see above)

Bestra replied:
@egh Is there a way to alias namespaces?  There are a few places that we're matching on `mc:` (`mc:Choice`) items, and this was the only way we knew of a the time to make the parser happy.

egh replied:
You can just change `mc:Choice` to `ve:Choice`. In a proper XML parser the prefix doesn't matter, only the URI that it maps to.

Bestra replied:
Really! That seems worth doing.

On Wed, Jul 27, 2016 at 12:36 PM, Erik Hetzner notifications@github.com
wrote:

> In docx/from/docxtotei.xsl
> https://github.com/Tahi-project/Stylesheets/pull/8#discussion_r72474816:
> 
> > -  xmlns:mml="http://www.w3.org/1998/Math/MathML"
> > -  xmlns:mo="http://schemas.microsoft.com/office/mac/office/2008/main"
> > -  xmlns:mv="urn:schemas-microsoft-com:mac:vml"
> > -  xmlns:o="urn:schemas-microsoft-com:office:office"
> > -  xmlns:pic="http://schemas.openxmlformats.org/drawingml/2006/picture"
> > -  xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"
> > -  xmlns:rel="http://schemas.openxmlformats.org/package/2006/relationships"
> > -  xmlns:tbx="http://www.lisa.org/TBX-Specification.33.0.html"
> > -  xmlns:html="http://www.w3.org/1999/xhtml"
> > -  xmlns:tei="http://www.tei-c.org/ns/1.0"
> > -  xmlns:teidocx="http://www.tei-c.org/ns/teidocx/1.0"
> > -  xmlns:v="urn:schemas-microsoft-com:vml"
> > -  xmlns:ve="http://schemas.openxmlformats.org/markup-compatibility/2006"
> > -  xmlns:w10="urn:schemas-microsoft-com:office:word"
> > -  xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main"
> > -  xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
> 
> You can just change mc:Choice to ve:Choice. In a proper XML parser the
> prefix doesn't matter, only the URI that it maps to.
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> https://github.com/Tahi-project/Stylesheets/pull/8/files/8feacc95c00616da21e45ddb6c282803e66eb995#r72474816,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AB8t1KrBHNRa0xW8LmO-dElLKDS4fL8Iks5qZ4j_gaJpZM4JMeq0
> .

MMercieca replied:
It seems to work locally so I pushed it up.  Let me know if it looks good @egh and @Bestra

egh commented on docx/from/docxtotei.xsl:
```diff
@@ -1,713 +1,750 @@
+
+
+    <!-- Save the result to a temp file for debugging purposes
+     <xsl:result-document href="/tmp/foo.xml">
```
Should we remove this?

Bestra replied:
Probably.  It was in the original file as a helpful example, but I think @MMercieca tried it and it didn't actually work :P .

MMercieca replied:
Yes, we should recomment that.  It was working, just not how I expected.

MMercieca replied:
It's in the source file and somewhat useful. I could remove it though if it helps make things cleaner.

egh replied:
Ah, cool. Makes sense to keep it then. It was hard to tell that it was there originally with all the whitespace changes.

egh commented on docx/from/docxtotei.xsl:
```diff
@@ -1,713 +1,750 @@
+    </xsl:template>
+
+    <xsl:template name="removeNestedTables" match="w:txbxContent/w:tbl" mode="preprocessing">
+      <w:p w:rsidR="009C0DAE" w:rsidRDefault="009C0DAE">
```
Are these identifiers meaningful?

Bestra replied:
If you're referring to the `w:rsid*` elements, then I don't think so. Those
are revision identifiers, I believe.
https://blogs.msdn.microsoft.com/brian_jones/2006/12/11/whats-up-with-all-those-rsids/

On Wed, Jul 20, 2016 at 12:10 PM, Erik Hetzner notifications@github.com
wrote:

> In docx/from/docxtotei.xsl
> https://github.com/Tahi-project/Stylesheets/pull/8#discussion_r71555784:
> 
> > -      xsl:choose
> > -        <xsl:when test="@w:name='_GoBack'"/>
> > -        <xsl:when test="starts-with(@w:name,'_')">
> > -          <xsl:copy-of select="."/>
> > -        /xsl:when
> > -        <xsl:when test="self::w:bookmarkEnd"/>
> > -        xsl:otherwise
> > -          <ANCHOR>
> > -            <xsl:attribute name="xml:id" select="@w:name"/>
> > -          </ANCHOR>
> > -        /xsl:otherwise
> > -      /xsl:choose
> > -    /xsl:template
> >   +
> > -    <xsl:template name="removeNestedTables" match="w:txbxContent/w:tbl" mode="preprocessing">
> > -      <w:p w:rsidR="009C0DAE" w:rsidRDefault="009C0DAE">
> 
> Are these identifiers meaningful?
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> https://github.com/Tahi-project/Stylesheets/pull/8/files/8feacc95c00616da21e45ddb6c282803e66eb995#r71555784,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AB8t1L_AkZvwJSrpPI9Ap9q5KZrq_L9lks5qXkhogaJpZM4JMeq0
> .

MMercieca replied:
Yup. I removed the revision identifiers.

Bestra commented:
@egh I'm good to merge this if you are.

egh commented:
Looks good to me. Whether or not you want to change the namespace duplication is up to you. :)

MMercieca committed:
APERATA-7046 Using ve namespace instead of mc
